File: Decorator/Decorator.ts
interface ICoffee {
  getDescription(): string;
  getCost(): number;
}

// Implementação concreta do café simples
class SimpleCoffee implements ICoffee {
  public getDescription(): string {
    return "Café simples";
  }

  public getCost(): number {
    return 1.0; // Preço do café simples
  }
}

// Decorator base
abstract class CoffeeDecorator implements ICoffee {
  protected coffee: ICoffee;

  constructor(coffee: ICoffee) {
    this.coffee = coffee;
  }

  public getDescription(): string {
    return this.coffee.getDescription();
  }

  public getCost(): number {
    return this.coffee.getCost();
  }
}

// Decorator para adicionar leite ao café
class MilkDecorator extends CoffeeDecorator {
  constructor(coffee: ICoffee) {
    super(coffee);
  }

  public getDescription(): string {
    return `${this.coffee.getDescription()}, com leite`;
  }

  public getCost(): number {
    return this.coffee.getCost() + 0.5; // Adiciona o custo do leite
  }
}

// Decorator para adicionar caramelo ao café
class CaramelDecorator extends CoffeeDecorator {
  constructor(coffee: ICoffee) {
    super(coffee);
  }

  public getDescription(): string {
    return `${this.coffee.getDescription()}, com caramelo`;
  }

  public getCost(): number {
    return this.coffee.getCost() + 0.75; // Adiciona o custo do caramelo
  }
}

// Uso do padrão Decorator para criar e manipular cafés decorados
let coffee: ICoffee = new SimpleCoffee();
console.log(coffee.getDescription()); // Saída: Café simples
console.log(`Custo: $${coffee.getCost()}`); // Saída: Custo: $1.0

coffee = new MilkDecorator(coffee);
console.log(coffee.getDescription()); // Saída: Café simples, com leite
console.log(`Custo: $${coffee.getCost()}`); // Saída: Custo: $1.5

coffee = new CaramelDecorator(coffee);
console.log(coffee.getDescription()); // Saída: Café simples, com leite, com caramelo
console.log(`Custo: $${coffee.getCost()}`); // Saída: Custo: $2.25

File: Decorator/Decorator.cs
// Café simples
ICoffee coffee = new SimpleCoffee();
Console.WriteLine(coffee.GetDescription()); // Saída: Café simples
Console.WriteLine($"Custo: ${coffee.GetCost()}"); // Saída: Custo: $1.0

// Café com leite
coffee = new MilkDecorator(coffee);
Console.WriteLine(coffee.GetDescription()); // Saída: Café simples, com leite
Console.WriteLine($"Custo: ${coffee.GetCost()}"); // Saída: Custo: $1.5

// Café com leite e caramelo
coffee = new CaramelDecorator(coffee);
Console.WriteLine(coffee.GetDescription()); // Saída: Café simples, com leite, com caramelo
Console.WriteLine($"Custo: ${coffee.GetCost()}"); // Saída: Custo: $2.25

interface ICoffee
{
  string GetDescription();
  decimal GetCost();
}

// Implementação concreta do café simples
class SimpleCoffee : ICoffee
{
  public string GetDescription()
  {
    return "Café simples";
  }

  public decimal GetCost()
  {
    return 1.0m; // Preço do café simples
  }
}

// Decorator base
abstract class CoffeeDecorator : ICoffee
{
  protected ICoffee _coffee;

  public CoffeeDecorator(ICoffee coffee)
  {
    _coffee = coffee;
  }

  public virtual string GetDescription()
  {
    return _coffee.GetDescription();
  }

  public virtual decimal GetCost()
  {
    return _coffee.GetCost();
  }
}

// Decorator para adicionar leite ao café
class MilkDecorator : CoffeeDecorator
{
  public MilkDecorator(ICoffee coffee) : base(coffee) { }

  public override string GetDescription()
  {
    return $"{_coffee.GetDescription()}, com leite";
  }

  public override decimal GetCost()
  {
    return _coffee.GetCost() + 0.5m; // Adiciona o custo do leite
  }
}

// Decorator para adicionar caramelo ao café
class CaramelDecorator : CoffeeDecorator
{
  public CaramelDecorator(ICoffee coffee) : base(coffee) { }

  public override string GetDescription()
  {
    return $"{_coffee.GetDescription()}, com caramelo";
  }

  public override decimal GetCost()
  {
    return _coffee.GetCost() + 0.75m; // Adiciona o custo do caramelo
  }
}
